[Woo POS] Coupons: Enable site coupons when tap on CTA

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15348 

## Description
This PR adds the ability of enabling the site's coupons when we tap the CTA. At the moment we're still using the full-screen error view to handle loading errors for products and coupons, but from the latests designs it seems we'll be moving to a partial screen (pdfdoF-6PX#comment-7945-p2) so I focused on the functionality and reusing the error view over the UI details.

https://github.com/user-attachments/assets/9c6c752b-8179-4a8a-9e6c-13e532eeac58

## Testing information
- Disable coupons in your site
- Load POS, navigate to coupons
- Tap on `Enable`, wait a moment (loading state to be handled)
- Observe Coupons being loaded

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Add error type to POS error state

iamgabrielma committed:
adapt error view action to accept different callbacks

iamgabrielma committed:
enable coupons when tap on cta

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/241">22.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
refactor for simplicity

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -163,6 +167,28 @@ extension EnvironmentValues {
 
+@available(iOS 17.0, *)
+private extension PointOfSaleDashboardView {
+    func enableCoupons() async {
```
I didn't want to go too deep into where to add this function right now: Initially I wanted to add it to the coupons service but this means that if we want to trigger it from the dashboard it has to cross through aggregate model -> coupon controller -> coupon service. Since it's a store-wide setting, would it make sense to just add it to the aggregate model or the coupon controller?

staskus replied:
Understandable issue. I think one place it doesn't belong in is `PointOfSaleDashboardView`.

> Since it's a store-wide setting, would it make sense to just add it to the aggregate model or the coupon controller?

I agree that it's store-wide, so the aggregate model could work. However, I think we don't want to make it have too many responsibilities.

Since we have a coupon logic-related controller, moving it at least to a controller makes more sense.

iamgabrielma committed:
make test compile

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -163,6 +167,28 @@ extension EnvironmentValues {
 
+@available(iOS 17.0, *)
+private extension PointOfSaleDashboardView {
+    func enableCoupons() async {
```
One more comment regarding that: when the button is tapped, we should make sure the item state reflects that and the view doesn't keep being stale. So it should switch to a loading state.

staskus reviewed (commented):
Thanks!

I left the logic about the dashboard view state.

Could you create the remaining tasks in Linear so we wouldn't miss any details to properly implement error, empty, and disabled states according to the designs and with the localized copy?

staskus commented:
@iamgabrielma I suggest checking, reviewing, testing, and merging [[Woo POS] Restore checkStoreCouponSettings to PointOfSaleCouponService](https://github.com/woocommerce/woocommerce-ios/pull/15481) before proceeding to avoid additional conflicts.

Also worth considering consistent place to call settings. We do it within `PointOfSaleCouponService` so we could continue doing it as well for enabling them 👍 .

iamgabrielma commented:
> I suggest checking, reviewing, testing, and merging https://github.com/woocommerce/woocommerce-ios/pull/15481 before proceeding to avoid additional conflicts.

Sounds good, I'll do that first as otherwise working on this one will become messy, then apply your suggestions.

> Could you create the remaining tasks in Linear so we wouldn't miss any details to properly implement error, empty, and disabled states according to the designs and with the localized copy?

Sure, I see you've already created these, I've added WOOMOB-253 so we don't forget to localize, or I'll do it just here when addressing the rest of changes.

iamgabrielma committed:
Merge branch 'trunk' into task/15348-followup-enable-coupon-via-cta

iamgabrielma committed:
moves enable coupons call to coupons controller

iamgabrielma committed:
move enable coupons to service

iamgabrielma committed:
add follow-up issue reference

iamgabrielma committed:
remove unnecessary import

iamgabrielma committed:
localizations

iamgabrielma commented:
This is ready for another pass when you have the chance @staskus 
-  I've moved the call to settings to the coupons service: https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/a2283f21c669864cb6044477a2ea0c49ff8ced9b
- Added an internal comment to follow-up with loading state https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/0ffcc1072144a2a4cb451a165adbb9e6c6c6bea8
- Added missing localizations to errors, we can edit them once we have the final copy: https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/fac5a0715801707fc32f69fddf276200b8e48461

iamgabrielma committed:
make test compile

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -21,9 +21,16 @@ protocol PointOfSaleItemsControllerProtocol {
 
-
+@available(iOS 17.0, *)
+extension PointOfSaleItemsControllerProtocol {
```
Do we need this extension with default implementation?

iamgabrielma replied:
We do unfortunately, since we need to call `couponsController.enableCoupons()` in the aggregate, and this is a specialization of couponsController `PointOfSaleItemsControllerProtocol`

staskus replied:
Oh, I didn't realize that at first. I don't love that `ItemController` gets affected this way. 

This is what we could do:

1. Create `PointOfSaleCouponsControllerProtocol` which extends `PointOfSaleItemsControllerProtocol`:

```
protocol PointOfSaleCouponsControllerProtocol: PointOfSaleItemsControllerProtocol {
    /// Enables coupons in store settings, if needed
    func enableCoupons() async
}
```

2. Implement it `final class PointOfSaleCouponsController: PointOfSaleCouponsControllerProtocol`

3. Use it in aggregate model: `private let couponsController: PointOfSaleCouponsControllerProtocol`

4. Profit

This way coupons controller can still share common funtionality with items controller but also grow with additional coupons related functionality.

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -68,6 +69,20 @@ public final class PointOfSaleCouponService: PointOfSaleCouponServiceProtocol {
+        _ = await withCheckedContinuation { continuation in
+            settingsStoreMethods.enableCouponSetting(siteID: siteID) { result in
+                switch result {
+                case .success(let success):
```
```suggestion
                case .success:
```

iamgabrielma replied:
Updated https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/093fc72560c2fc56deada201f7ab946fb2e3e522

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -68,6 +69,20 @@ public final class PointOfSaleCouponService: PointOfSaleCouponServiceProtocol {
+                switch result {
+                case .success(let success):
+                    continuation.resume(returning: true)
+                case .failure(let failure):
```
```suggestion
                case .failure:
```

iamgabrielma replied:
Updated https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/093fc72560c2fc56deada201f7ab946fb2e3e522

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -68,6 +69,20 @@ public final class PointOfSaleCouponService: PointOfSaleCouponServiceProtocol {
     }
+
+    @MainActor
+    public func enableCoupons() async {
```
Should it throw an error so the controller could set some sort of `enableCouponErrorState` with an option to retry enabling it?

```suggestion
    public func enableCoupons() async throws {
```

iamgabrielma replied:
Yes, this is tracked by WOOMOB-255 but we can already change the signature of this to be async throws: https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/fb126bc37fc221c755e838f16faf849f0a07b5a0

staskus replied:
WOOMOB-255 doesn't mention error handling. If it's going to be handled there we should update the task.

iamgabrielma replied:
Oh, you're totally right. I've logged this in a new WOOMOB-267. We might be able to close both issues in the same PR, but I'll keep the tasks separate for now as are more approachable.

staskus reviewed (commented):
Thanks! Looks good 👍 

It would be good to add a couple of missing pieces so the solution would be full and only UI missing:

1. Handle in controller when service fails to enable coupons
2. Add tests

Thanks 👍

iamgabrielma committed:
Merge branch 'trunk' into task/15348-followup-enable-coupon-via-cta

# Conflicts:
#	WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift
#	WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift

iamgabrielma committed:
make compile. ammend merge conflict

iamgabrielma commented:
👋 I'll need to iterate on the approach we took here, as with the changes on the controller and item view state we've recently merged this use case doesn't work anymore: 

From what I've checked so far, when switching to coupons, and coupons throw the `.errorCouponsDisabled` error on loading items,  the `ItemsViewState` we set does not seem to longer render the correct case, as the `PointOfSaleDashboardView` does not switch again between cases when performing `switch posModel.itemsViewState.containerState {`, so we keep loading the previous `contentView` case which doesn't display anything as coupons are 0 when disabled. I'll look into this today.

iamgabrielma commented:
> switch posModel.itemsViewState.containerState 

Issue seems to be here, since we do not use a "current controller", when we render the dashboard we're only checking for changes in the "item controller" which not really an item controller but a product controller, changes on `couponsViewState` are not handled so views are not re-rendered.

iamgabrielma committed:
Add a shared current item view state based on item type

When view state changed, driven by changes in coupons, these were never propagated to the dashboard view since we only switched in “itemsViewState” which only covered the products controller.

iamgabrielma committed:
remove unnecessary associated values

iamgabrielma committed:
update enableCoupons signature to be throwable

iamgabrielma commented:
Thanks for the review again @staskus  

It seems we still need to expose some sort of "current" controller or view state to the views from the aggregate, otherwise we were only publishing changes in the "product view state", not on the coupons view state.I've added this on https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/562c1c14a828e289d84a297e116061009e7f610d

I've also added some minimal renaming for "items", and named them "products". Otherwise gets very confusing to work with these lists and state and be sure what its actually updating, and one of the reasons we missed this issue originally (we're calling "items" to products, and calling "items" to items). I think we still need to rename this further across the board, but we can do this separately.

Xcode has decided today that doesn't want to compile the tests locally (I've cleared derived data a few times, and rebuilt, but is not very useful) but they're passing CI. If you're ok with this I'll add the tests separately, for now I'll leave this PR for review to move the ball forward as I'm EOD:

<img width="270" alt="Screenshot 2025-04-07 at 16 15 26" src="https://github.com/user-attachments/assets/583a2588-9a5d-4766-81ee-13abe528199c" />

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -47,15 +50,15 @@ struct PointOfSaleDashboardView: View {
             .trackSize(size: $floatingSize)
             .accessibilitySortPriority(1)
-            .renderedIf(posModel.itemsViewState.containerState != .loading)
+            .renderedIf(posModel.currentViewState.containerState != .loading)
```
This took me a while: By leaving this one as `posModel.itemViewstate`, when switching to Coupons we were always stuck in .loading state, not showing coupons correctly. Of course the compiler didn't complain, but was tricky to find why the view state was wrong. I think we can move those ones to some computed property, or grab them directly from reading a variable from the aggregate model.

staskus replied:
In general, this containerState logic doesn't work for products and coupons anymore, since we don't want to be showing coupons errors in full screen. This structure may change 🤔

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -24,7 +24,8 @@ protocol PointOfSaleAggregateModelProtocol {
     func trackCardPaymentsOnboardingShown()
 
-    var itemsViewState: ItemsViewState { get }
+    var productsViewState: ItemsViewState { get }
```
I encourage not to make this renaming unless we rename all the items to products in the codebase. 

As far as our internal codebase and business case knowledge goes - items include products and product variations

iamgabrielma replied:
Gotcha, I've reverted it on https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/e1979c9eaa8f159780486885a34f7159288b2b90 I still think we should rename this at some point across the codebase as coupons should be considered items as well in certain layers.

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -34,7 +43,7 @@ protocol PointOfSaleSearchingItemsControllerProtocol: PointOfSaleItemsController
 
 @available(iOS 17.0, *)
-@Observable final class PointOfSaleItemsController: PointOfSaleSearchingItemsControllerProtocol {
+@Observable final class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol, PointOfSaleSearchingItemsControllerProtocol {
```
`PointOfSaleSearchingItemsControllerProtocol` extends `PointOfSaleItemsControllerProtocol`, no need to add it again.

iamgabrielma replied:
Fixed! https://github.com/woocommerce/woocommerce-ios/pull/15478/commits/f7db770c54c9af7d714c0c62a026c553683f0fee

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -98,22 +103,33 @@ protocol PointOfSaleAggregateModelProtocol {
 @available(iOS 17.0, *)
 extension PointOfSaleAggregateModel {
+    func updateCurrentViewState(base: ItemListBaseItem) {
+        let viewState = base.itemType == .products ? productsViewState : couponsViewState
```
If `currentViewState` is required, I wonder if we should approach it differently.

`ItemListView` already holds `selectedItemType`. Maybe we should track selection in a single place—an aggregate model? That would give a source of truth and allows to keep `currentViewState` and even `currentController` as computed helper variables. 

`ItemListView` also doesn't need to switch for `itemsStack` anymore and can use `posModel.currentViewState.itemsStack`.

Wdyt?

iamgabrielma replied:
Yes, I think so. This also should help to keep proper state on which tabs we should be loading (products v coupons) when something from outside the list view happens (eg: errors, retry, etc, ... ). I've logged this on  WOOMOB-269

staskus reviewed (commented):
Thank you! I understand the need for `currentViewState` 👍  , I think we could go this way but leave only one source of truth. 

Also, I noticed that when coupons enabled, products tab is opened. Coupons tab should still be selected:


https://github.com/user-attachments/assets/af3d3e42-1280-4e52-9c6a-e527dd7f3ae7

iamgabrielma committed:
update issue references

iamgabrielma committed:
revert itemsViewState rename

iamgabrielma committed:
remove duplicated conformance

iamgabrielma committed:
make PointOfSaleCouponsControllerProtocol

iamgabrielma committed:
update tests with new mock

iamgabrielma committed:
lint

iamgabrielma commented:
I've split the feedback between changes I'm doing in this one vs changes we can follow-up with on different issues, so I can scope this one, as the protocol changes come with updates to ~30 files and can get messy:

Changes:
- Reverted renaming
- Removed duplicated conformance
- Made a new protocol and mock/preview for coupons

To do:
- Additional tests will come on a follow-up PR
- Error handling from failing to enable settings is logged on WOOMOB-267
- Selecting the coupons tab after enabling it successfully is logged on WOOMOB-268. I've time-boxed 30 minutes here but I couldn't figure out why it's behaving like this. We only seem to toggle the tab state internally via `displayItemType()`. That said, this could be resolved by itself when we move from full-screen to side-screen.
- Centralizing the view state into a single source of truth is logged on WOOMOB-269

iamgabrielma commented:
> I noticed that when coupons enabled, products tab is opened. Coupons tab should still be selected.

Just to add: At some point it worked well, but then I could replicate the issue every time I've tried. I think once we make further changes regarding handling the view correctly (not full-screen) and move the source of truth of selected item somewhere else it should be either resolved or easier to resolve. For now is logged on WOOMOB-268

https://github.com/user-attachments/assets/bfa82d87-a9bc-4d68-bed5-fc9667ef952c

staskus reviewed (approved):
Thank you for the fixes and creating additional tasks. 👍 

I think we took this task too early, or tried to achieve too much with this PR. We really needed an infrastructure of showing a state within the list to properly tackle this. Now, we have to merge half-baked happy and unhappy paths.
